ledger/simulation/resources.go

```diff
@@ -605,3 +607,690 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
+			err = p.txnResources[i].addBox(app, name)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// If we have gotten to this point and the App ID is 0, then we know there is no where it can go
+	// The next loop will try to add the app reference, which is nonsensical for 0
+	if app == basics.AppIndex(0) {
+		return err
+	}
+
+	// Then try to find txn with room for both app and box
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addBoxWithApp(app, name)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) addHolding(addr basics.Address, aid basics.AssetIndex) error {
+	var err error
+
+	// First try to find txn with account already available that can also accommodate the asset.
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasAccount(addr) {
+			err = p.txnResources[i].addAsset(aid)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Then try to find txn with asset already available that can also accommodate the account.
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasAsset(aid) {
+			err = p.txnResources[i].addAccount(addr)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Finally try to find txn with room for both account and holding
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAssetHolding(addr, aid)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) addLocal(addr basics.Address, aid basics.AppIndex) error {
+	var err error
+
+	// First try to find txn with account already available	that can also accommodate the app.
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasAccount(addr) {
+			err = p.txnResources[i].addApp(aid)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Then try to find txn with app already available that can also accommodate the account.
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasApp(aid) {
+			err = p.txnResources[i].addAccount(addr)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Finally try to find txn with room for both account and app
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAppLocal(aid, addr)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) populateResources(groupResourceTracker ResourceTracker, txnResources []ResourceTracker) error {
+	// We don't want to mutate the groupResourceTracker because it is used later in simulate for UnnamedResourcesAccessed
+	groupResources := struct {
+		Assets        []basics.AssetIndex
+		Apps          []basics.AppIndex
+		Accounts      []basics.Address
+		Boxes         []logic.BoxRef
+		AssetHoldings []ledgercore.AccountAsset
+		AppLocals     []ledgercore.AccountApp
+	}{
+		Apps:          []basics.AppIndex{},
+		Accounts:      []basics.Address{},
+		Boxes:         []logic.BoxRef{},
+		AssetHoldings: []ledgercore.AccountAsset{},
+		AppLocals:     []ledgercore.AccountApp{},
+	}
```

Comment:
Why is this anonymous struct type introduced? I think you just have 6 variables here that never get passed together anywhere, and no operations on this type.